File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # approach: sliding window, set of chars in current window
        
        if not s:
            return 0
        
        result = 0
        window = set()
        
        l = 0
        for r, c in enumerate(s):
            while c in window:
                window.remove(s[l])
                l += 1
            
            window.add(c)
            result = max(result, len(window))
        
        return result
            

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // approach: hashset to store nums in current valid window
        
        int result = 0;
        int l = 0;
        
        Set<Character> window = new HashSet<>();
        
        for (int r = 0; r < s.length(); r++) {
            char c = s.charAt(r);
            
            while (window.contains(c)) {
                window.remove(s.charAt(l));
                l++;
            }
            
            window.add(c);
            result = Math.max(result, window.size());
        }
        
        return result;
    }
}